publish genesis commit on ledger creation

dpetran opened the pull request:
Now a ledger will `exist?` once created.

dpetran committed:
publish genesis commit on ledger creation

Now a ledger will `exist?` once created.

dpetran committed:
fix cljs warnings

dpetran committed:
fix address split on non-jvm platforms

For some reason the regex #":(?!.*:)" was producing a different result on the browser
and in nodejs than on the jvm:

; clj:
(split-address "fluree:memory://testledger")
["fluree:memory" "//testledger"]

; cljs:
(split-address "fluree:memory://testledger")
["fluree" "memory://testledger"]

All the documentation I could find indicates that this regex should in fact work
cross-platform, but for some reason it reliably did not. I have replaced its use with an
alternate implementation that safely avoids the issue.

zonotope commented on src/clj/fluree/db/ledger.cljc:
```diff
@@ -163,6 +165,16 @@
     (subs ledger-alias 1)
     ledger-alias))
 
+(defn publish-commit
```
This function operates on a connection, so it should be in the connection namespace. Ledgers shouldn't reach in to connections, and they should not know anything about name services, or anything else attached to a connection besides the ledger itself. Publishing commits is not and should not be the concern of ledger objects.

dpetran replied:
Good catch, I've refactored to do the correct thing instead of the expedient thing :+1:

zonotope commented on src/clj/fluree/db/storage.cljc:
```diff
@@ -48,7 +48,8 @@
   [address]
-  (str/split address #":(?!.*:)" 2))
+  (let [components (str/split address #":")]
+    [(str/join ":" (butlast components)) (last components)]))
```
was there a bug how it was previously? this looks like it's doing a bunch of extra work to split the entire string only to immediately join it again.

dpetran replied:
Yes, I've detailed the problem in the commit message of [ab50edc](https://github.com/fluree/db/pull/915/commits/ab50edcb7715b10dcc5c62e5b733bc3075fbfc9d). I've update this implementation to be more efficient.

dpetran committed:
make split-address more efficient

dpetran committed:
move publish-commit back to connection

ledger/create is just a thin wrapper around ledger/instantiate, so in order to preserve
ledger's ignorance of nameservices and connection internals I moved the basic setup to
the connection. This has the nice side effect of removing the need for the internal-only
opt for instantiating init-time during migration, since the call site is now responsible
for declaring the init time.

bplatz commented:
I still have a philosophical issue with the genesis block.

We've tried to make dbs 'content addressed', and therefore without content, there is nothing to address (except commit metadata - but that isn't a db).

Aliases should be mappable to any "db" - but that is controlled by the nameservice(s) you are using - which might also be your own local nameservice (just like you can locally modify DNS with `/etc/hosts`)

Granted we are working with what we've got, so my input here might be idealistic and less practical for right now, but my preference would be to:
a) register/unregister/update aliases on the nameservice. The alias can be part of the commit, but is just a 'suggestion'.
b) There is no genesis commit, only a first commit of data which is content-addressed

zonotope commented:
> I still have a philosophical issue with the genesis block.
> 
> We've tried to make dbs 'content addressed', and therefore without content, there is nothing to address (except commit metadata - but that isn't a db).
> 
> Aliases should be mappable to any "db" - but that is controlled by the nameservice(s) you are using - which might also be your own local nameservice (just like you can locally modify DNS with `/etc/hosts`)
> 
> Granted we are working with what we've got, so my input here might be idealistic and less practical for right now, but my preference would be to: a) register/unregister/update aliases on the nameservice. The alias can be part of the commit, but is just a 'suggestion'. b) There is no genesis commit, only a first commit of data which is content-addressed

I think in order for this to work, we'd have to remove the alias and branch from the commit document and instead associate it with a name service. I think we'd also need to introduce hierarchical name services too as right now name services are flat, and a connection round-robin checks each one. So, instead of having a local file name service sitting along side an ipfs name service that sits lower in the priority (for example) we would have a file name service that referenced the ipfs name service and looked up addresses after doing whatever remapping it would need to do. 

I don't think I disagree with any of this, but I'm just trying to work through the implications and what we'd need to do to make it work. In any case, I don't think we should keep what we have now because the `create-with-transaction` flow is both confusing for users and has lead to bugs in implementation (you can `create` a ledger with the db library but not when using that db library in a stateless context like in fluree/server, so an error during the first create could lead to other errors about a phantom ledger already existing).

dpetran committed:
stop publishing the genesis commit

bplatz commented:
> I would like to make a case for it, though, on the philosophical level : )

I agree with the case - I disagree on how we enable it is all. I think how you create a ledger is you reserve the name in the desired authoritative nameservice. Now nobody else can use that name or create it using that name, and presumably nobody else can publish commit updates to that name that don't have authorization.

We talked about the nameservice being managed via a "system" Fluree ledger... this could also open up all the goodness around policy enforcement and time travel that Fluree offers. The Fluree ledger then also has the controls for guaranteed ordering.

Ultimately that's what I'd like to see... and it not only solves this problem it solves other problems.

While unlikely, I also think your proposed approach doesn't have thread safety... whereas a Fluree ledger solves that for you. If two parties request the identical ledger name at the identical moment in time both would incorrectly succeed I believe - unless those are running through consensus (so this wouldn't be a problem if using Raft).

Ben and I see this problem a little differently I think - so if he and you agree then I'm happy to have the proposed changes you suggest. I think this other approach however will create a better system - we just don't have all the pieces in place for it yet.

dpetran commented:
Ok, that clarifies a lot, thanks. It'll take some thought to figure out the path from here to there, but I don't think I object to your approach.

dpetran commented:
Closing for now, cljs fixes here: https://github.com/fluree/db/pull/977
